lib/graphql/metrics/tracer.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+module GraphQL
+  module Metrics
+    class Tracer
+      # NOTE: Used to store timings from lexing, parsing, validation, before we have a context to store
+      # values in. Uses thread-safe Concurrent::ThreadLocalVar to store a set of values per thread.
+
+      # NOTE: These constants come from the graphql ruby gem.
+      GRAPHQL_GEM_LEXING_KEY = 'lex'
+      GRAPHQL_GEM_PARSING_KEY = 'parse'
+      GRAPHQL_GEM_VALIDATION_KEYS = ['validate', 'analyze_query']
+      GRAPHQL_GEM_TRACING_FIELD_KEYS = [
+        GRAPHQL_GEM_TRACING_FIELD_KEY = 'execute_field',
+        GRAPHQL_GEM_TRACING_LAZY_FIELD_KEY = 'execute_field_lazy'
+      ]
+
+      def trace(key, data)
+        # NOTE: Context doesn't exist yet during lexing, parsing.
+        possible_context = data[:query]&.context
+
+        skip_tracing = possible_context&.fetch(GraphQL::Metrics::SKIP_GRAPHQL_METRICS_ANALYSIS, false)
+        return yield if skip_tracing
+
+        # NOTE: Not all tracing events are handled here, but those that are are handled in this case statement in
+        # chronological order.
+        case key
+        when GRAPHQL_GEM_LEXING_KEY
+          return setup_tracing_before_lexing { yield }
```

Comment:
I thought we had a conversation about just passing the block directly instead of `{ yield }`, did this actually end up being the faster approach?